Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift

```diff
@@ -1232,11 +1232,11 @@ internal final class PubGrubPackageContainer {
             }
         }
 
-        let (lowerBounds, upperBounds) = try self.computeBounds(
+        /*let (lowerBounds, upperBounds) = try self.computeBounds(
             for: node,
             constraints: constraints,
             startingWith: version
-        )
+        )*/
```

Comment:
computation of the bounds here is _very_ expensive and it needs to evaluate manifests throughout the dependency closure. afaict this is only used for the `emittedIncompatibilities` which is itself a performance optimization. disabling this gave 3x performance improvement on my initial tests

Reply:
cc @SDGGiesbrecht as this reflects our previous discussion about this optimization

Reply:
🤷 If it’s not actually faster, it’s not actually faster.

If you remove this, then `emittedIncompatibilities` is never written to and thus always empty, which in turn means...
- The line with `if emittedIncompatibilities[node]?.contains(version) != true {` is always true, and the contents of the `if` may as well be run unconditionally.
- The line with `emittedIncompatibilities.isEmpty` is also always true, although not necessarily the other conditions grouped with it.

While looking at it, I also noticed that [this section](https://github.com/apple/swift-package-manager/blob/15d19824817917b7faf7680c5a2d8a4bd459c16e/Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift#L1219-L1232) and [this section](https://github.com/apple/swift-package-manager/blob/15d19824817917b7faf7680c5a2d8a4bd459c16e/Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift#L1241-L1266) are now basically identical—comments and all. So you can probably reduce [this](https://github.com/apple/swift-package-manager/blob/15d19824817917b7faf7680c5a2d8a4bd459c16e/Sources/PackageGraph/Pubgrub/PubgrubDependencyResolver.swift#L1207-L1233) to just the following and let it naturally fall through into the other, identical section.

```swift
if version == pinnedVersion {
  if self.emittedPinnedVersionIncompatibilities[node] ?? false { return [] }
  self.emittedPinnedVersionIncompatibilities[node] = true
} 
```

It might be worth checking whether those remaining lines are actually helpful for performance. If not, you could get rid of `emittedPinnedVersionIncompatibilities` too.

Reply:
This particular optimization predates me. The [commit that added it](https://github.com/apple/swift-package-manager/commit/aee02eb246442bcc28119101b89feee1d54110b4) contains a bit of an explanation and a radar link (to which I lack access). It might be worth checking that radar for specific examples to double‐check for performance regressions. Also, if @aciidb0mb3r is still around out there somewhere, he might have something to add.

Reply:
Could it have an impact on the quality of the diagnostics when there are conflicts?

Reply:
we have a fairly robust set of tests for dependency resolution diagnostics which _have been_ impacted by #3985 but not by this. afaict the code here was added in https://github.com/apple/swift-package-manager/commit/aee02eb246442bcc28119101b89feee1d54110b4 as an attempt to improve diagnostics but because it is so expensive it is also cached. we already re-did the part that impacted the diagnostics in #3985 by taking a bit of a different approach and such this PR is purely about performance in not running that expensive calculation any more as it no longer has any functional value

Reply:
this is now https://github.com/apple/swift-package-manager/pull/4007, let take the discussion there?